src/mem/ruby/system/GPUCoalescer.cc

```diff
@@ -556,23 +556,46 @@ GPUCoalescer::hitCallback(CoalescedRequest* crequest,
     //
     // MUST AD DOING THIS FOR EACH REQUEST IN COALESCER
     std::vector<PacketPtr> pktList = crequest->getPackets();
+
+    uint8_t* log = nullptr;
     DPRINTF(GPUCoalescer, "Responding to %d packets for addr 0x%X\n",
             pktList.size(), request_line_address);
+    uint32_t offset;
+    int pkt_size;
     for (auto& pkt : pktList) {
-        request_address = pkt->getAddr();
+        offset = getOffset(pkt->getAddr());
+        pkt_size = pkt->getSize();
         if (pkt->getPtr<uint8_t>()) {
-            if ((type == RubyRequestType_LD) ||
-                (type == RubyRequestType_ATOMIC) ||
-                (type == RubyRequestType_ATOMIC_RETURN) ||
-                (type == RubyRequestType_IFETCH) ||
-                (type == RubyRequestType_RMW_Read) ||
-                (type == RubyRequestType_Locked_RMW_Read) ||
-                (type == RubyRequestType_Load_Linked)) {
-                pkt->setData(
-                    data.getData(getOffset(request_address), pkt->getSize()));
-            } else {
-                data.setData(pkt->getPtr<uint8_t>(),
-                             getOffset(request_address), pkt->getSize());
+            switch(type) {
+                case RubyRequestType_ST:
+                case RubyRequestType_ATOMIC_NO_RETURN:
+                    data.setData(pkt->getPtr<uint8_t>(), offset, pkt_size);
+                    break;
+                case RubyRequestType_LD:
+                    pkt->setData(data.getData(offset, pkt_size));
+                    break;
+                case RubyRequestType_ATOMIC_RETURN:
+                    if (pkt->isAtomicOp()) {
+                        // Atomic operations are performed by the WriteMask
+                        // in packet order, set by the crequest. Thus, when
+                        // unpacking the changes from the log, we read from
+                        // the front of the log to correctly map response
+                        // data into the packets.
+                        //
+                        // Log entry contains the old value before the current
+                        // atomic operation occurred.
+                        assert(data.numAtomicLogEntries());
+                        log = data.popAtomicLogEntryFront();
+                        assert(data.numAtomicLogEntries() >= 0);
+                        pkt->setData(&log[offset]);
+                        delete [] log;
+                        log = nullptr;
+                    }
```

Comment:
is there a case where the type is an atomic return but isAtomicOp returns false?

Reply:
I believe that that case should never happen, but I vaguely remember seeing this pop up on some of the tester benchmarks, but I'm failing to remember which ones. Will test this out and see if the if-check is necessary.

Reply:
Looks like I remembered wrong, and I did not see a case when the packet->isAtomicOp() needs after already seeing that the ruby request type is one of the atomic types. Put in some assert statements so that this assumption is clear moving forward.